chore: enhance debugging docs (#945)
defenseunicorns/pepr: docs/060_best-practices/README.md

## Debugging

- [Debugging During Module Development](https://docs.pepr.dev/main/best-practices/#debugging-during-module-development)
- [Logging](https://docs.pepr.dev/main/best-practices/#logging)
- [Internal Error Occurred](https://docs.pepr.dev/main/best-practices/#internal-error-occurred)
- [Pepr Store](https://docs.pepr.dev/main/best-practices/#pepr-store)

Welcome to the the debugging section! 🐛

Pepr is composed of `Modules` (ie, what happens when you issue `npx pepr init`), [Capabilities](https://docs.pepr.dev/main/user-guide/capabilities/) like `hello-pepr.ts`, and [Actions](https://docs.pepr.dev/main/user-guide/actions/) (ie, the blocks of code containing filters and `Mutate`, `Validate`, `Watch`, `Reconcile`, `OnSchedule`). You can have as many Capabilities as you would like in a Module.

> Pepr is a webhook-based system, meaning it is event-driven. When a resource is created, updated, or deleted, Pepr is called to perform the actions you have defined in your Capabilities. It's common for multiple webhooks to exist in a cluster, not just Pepr. When there are multiple webhooks, the order in which they are called is not guaranteed. The only garuantee is that all of the `MutatingWebhooks` will be called before all of the `ValidatingWebhooks`. After the admission webhooks are called, the `Watch` and `Reconcile` are called. The `Reconcile` and `Watch` create a watch on the resources specified in the `When` block and are watched for changes after admission. The difference between reconcile and watch is that `Reconcile` processes events in a queue to guarantee that the events are processed in order where as watch does not. 

Reviewer: ```suggestion
Pepr is a webhook-based system, meaning it is event-driven. When a resource is created, updated, or deleted, Pepr is called to perform the actions you have defined in your Capabilities. It's common for multiple webhooks to exist in a cluster, not just Pepr. When there are multiple webhooks, the order in which they are called is not guaranteed. The only guarantee is that all of the `MutatingWebhooks` will be called before all of the `ValidatingWebhooks`. After the admission webhooks are called, the `Watch` and `Reconcile` are called. The `Reconcile` and `Watch` create a watch on the resources specified in the `When` block and are watched for changes after admission. The difference between reconcile and watch is that `Reconcile` processes events in a queue to guarantee that the events are processed in order where as watch does not. 
```

Considering that many webhooks may be modifying the same resource, it is best practice to validate the resource after mutations are made to ensure that the resource is in a valid state if it has been changed since the last mutation.

**_It is always a best practice to add a `Validate` to your `Mutate` to ensure that the object is in a valid state before making changes._**

```typescript
When(a.Pod)
  .IsCreated()
  .InNamespace("my-app")
  .WithName("database")
  .Mutate(pod => {
    pod.metadata.labels["pepr"] = "true";
    return pod;
  })
  // another mutating webhook could removed labels

>   .Validate(pod => {
>     if (pod.metadata.labels["pepr"] !== "true") {
>       return ["Label 'pepr' must be 'true'"];
>     }

Author: ```suggestion
  .Validate(pod => {
    if (pod.metadata.labels["pepr"] !== "true") {
      return pod.Approve("Label 'pepr' must be 'true'");
    }
    return pod.Deny("Needs pepr label set to true")
```

This allows you to set breakpoints in `Mutate()`, `Validate()`, `Reconcile()`, `Watch()` or `OnSchedule()` and step through module code.

Note that you will need a cluster running: 

```bash
k3d cluster create pepr-dev --k3s-arg '--debug@server:0' --wait
```

```typescript
When(a.Pod)
  .IsCreated()
  .InNamespace("my-app")
  .WithName("database")
  .Mutate(pod => {
    // Set a breakpoint here
    pod.metadata.labels["pepr"] = "true";
    return pod;
  })
  .Validate(pod => {
    // Set a breakpoint here
    if (pod.metadata.labels["pepr"] !== "true") {
      return ["Label 'pepr' must be 'true'"];
    }
  });
```

### Logging

Pepr can deploy two types of controllers: Admission and Watch. The controllers deployed are dictated by the [Actions](https://docs.pepr.dev/main/user-guide/actions/) called for by a given set of Capabilities (Pepr only deploys what is necessary). Within those controllers, the default log level is `info` but that can be changed to `debug` by setting the `LOG_LEVEL` environment variable to `debug`.

To pull logs for all controller pods:

```bash

> kubectl logs -l app.kubernetes.io/controller

Author: ```suggestion
kubectl logs -l app -n pepr-system
```

```

#### Admission Controller

If the focus of the debug is on a `Mutate()` or `Validate()`, the relevenat logs will be from pods with label `pepr.dev/controller: admission`.

```bash
kubectl logs -l pepr.dev/controller=admission -n pepr-system
```

More refined admission logs -- which can be optionally filtered by the module UUID -- can be obtained with [`npx pepr monitor`](https://docs.pepr.dev/main/best-practices/#monitoring)

```bash
npx pepr monitor 
```

#### Watch Controller

> If the focus of the debug is a `Watch()`, `Reconcile()`, or `OnSchedule()`, look for logs from pods containing label `pepr.dev/controller: watch`.

Author: ```suggestion
If the focus of the debug is a `Watch()`, `Reconcile()`, or `OnSchedule()`, look for logs from pods containing label `pepr.dev/controller: watcher`.
```

```bash

> kubectl logs -l pepr.dev/controller=watch -n pepr-system

Author: ```suggestion
kubectl logs -l pepr.dev/controller=watcher -n pepr-system
```

Pepr is a webhook-based system, meaning it is event-driven. When a resource is created, updated, or deleted, Pepr is called to perform the actions you have defined in your Capabilities. It's common for multiple webhooks to exist in a cluster, not just Pepr. When there are multiple webhooks, the order in which they are called is not guaranteed. The only guarantee is that all of the `MutatingWebhooks` will be called before all of the `ValidatingWebhooks`. After the admission webhooks are called, the `Watch` and `Reconcile` are called. The `Reconcile` and `Watch` create a watch on the resources specified in the `When` block and are watched for changes after admission. The difference between reconcile and watch is that `Reconcile` processes events in a queue to guarantee that the events are processed in order where as watch does not. 

Considering that many webhooks may be modifying the same resource, it is best practice to validate the resource after mutations are made to ensure that the resource is in a valid state if it has been changed since the last mutation.

> **_It is always a best practice to add a `Validate` to your `Mutate` to ensure that the object is in a valid state before making changes._**

Author: ```suggestion
```

## Debugging

- [Breakpoints](https://docs.pepr.dev/main/best-practices/#using-breakpoints)
- [Logging](https://docs.pepr.dev/main/best-practices/#logging)
- [Internal Error Occurred](https://docs.pepr.dev/main/best-practices/#internal-error-occurred)
- [Pepr Store](https://docs.pepr.dev/main/best-practices/#pepr-store)

> _It is always a best practice to add a `Validate` to your `Mutate` to ensure that the object is in a valid state before making changes._

Reviewer: I think this might need some more explaining -- iIt's not obvious (at least to me?) how I'd go about "adding a Validate to my Mutate" at a code level, so... maybe some expansion / an example is in order?

I'm assuming that this is intended to say something like "If you add a Mutate action, you should also add a Validate action to be sure that 1) your Mutate runs successfully, and 2) your resource wasn't changed unexpected/unreasoably by _other_ Mutates before being submitted for Validation."

Something like that maybe.  Unless there's a special validate-within-mutates facility that I'm not aware of / or the risks & consequences are described elsewhere..? :man_shrugging:

Author: addressed in pr

> _If you think your Webhook is not being called for a given resource, check the `*WebhookConfiguration`._

Reviewer: I kind of expect troubleshooting docs to be one of the first places new folks'll go when they start having problems working with / understanding Pepr, so... it might be worth expanding a little here too..?

For example, a new-ish person might think: "I write 'Modules' that aggregate 'Capabilities' which contain 'Actions' in Pepr, don't I?  Where are 'Webhooks' in the Pepr?"

Do you think we could expand this like so that it's a bit easier to grok for non-k8s-native types?  How do lead them to reasonable, actionable links / troubleshooting activities?

Author: addressed in pr

> ### Using Breakpoints

Reviewer: ```suggestion
### Debugging During Module Development
```

"Using Breakpoints" is a bit too wide-open, IMO -- leaves too much room for interpretation / room to hope that you might be able to debug a production cluster, which... is not possible.  I think something like this helps scope expectation a bit better, but I'm open to other phrasings if you like them more.

Author: addressed in pr

> Pepr supports breakpoints in the VSCode editor. To use breakpoints, run `npx pepr dev` in the root of the Pepr module in a JavaScript Debug Terminal. This command starts the Pepr development server running at `localhost:3000` with the `*WebhookConfiguration` configured to send `AdmissionRequest` objects to the local address.

Reviewer: ```suggestion
Pepr supports breakpoints in the VSCode editor. To use breakpoints, run `npx pepr dev` in the root of a Pepr module using a JavaScript Debug Terminal. This command starts the Pepr development server running at `localhost:3000` with the `*WebhookConfiguration` configured to send `AdmissionRequest` objects to the local address.
```
Minor edits to not assume that "the Pepr module" has already been introduced.

Author: addressed in pr

> This allows you to set breakpoints in `Mutate()`, `Validate()`, `Reconcile()`, `Watch()` or `OnSchedule` and step through the code.

Reviewer: ```suggestion
This allows you to set breakpoints in `Mutate()`, `Validate()`, `Reconcile()`, `Watch()` or `OnSchedule()` and step through module code.
```
Normalizing parens.

Author: addressed in pr

Note that you will need a cluster running: 

```bash
k3d cluster create pepr-dev --k3s-arg '--debug@server:0' --wait
```

```typescript
When(a.Pod)
  .IsCreated()
  .InNamespace("my-app")
  .WithName("database")
  .Mutate(pod => {
    // Set a breakpoint here
    pod.metadata.labels["pepr"] = "true";
    return pod;
  })
  .Validate(pod => {
    // Set a breakpoint here
    if (pod.metadata.labels["pepr"] !== "true") {
      return ["Label 'pepr' must be 'true'"];
    }
  });
```

### Logging

> Pepr can deploy two controllers: Admission and Watch. The controllers deployed are dictated by the [actions](https://docs.pepr.dev/main/user-guide/actions/) used in the capabilities (Pepr only deploys what is necessary). The default log level is `info`, but it can be changed to `debug` by setting the environment variable `LOG_LEVEL` to `debug`.

Reviewer: ```suggestion
Pepr can deploy two types of controllers: Admission and Watch. The controllers deployed are dictated by the [Actions](https://docs.pepr.dev/main/user-guide/actions/) called for by a given set of Capabilities (Pepr only deploys what is necessary). Within those controllers, the default log level is `info` but that can be changed to `debug` by setting the `LOG_LEVEL` environment variable to `debug`.
```

> Pepr logs for all pods:

Reviewer: ```suggestion
To pull logs for all controller pods:
```

```bash

> kubectl logs -l app.kubernetes.io/controller

Reviewer: This looks like it'll pull logs for all controllers... which is pretty broad, no?  Is there really no way to scope the label to Pepr controllers (of both types) only?  I could swear I remember some label with "pepr.dev" in it but... maybe I'm misremembering.

`kubectl logs -l 'pepr.dev/controller in (admission, watch)'` maybe?  Maybe the `-n pepr-system` flag would be helpful here too / otherwise (like the examples below)?

```

#### Admission Controller

If the focus of the debug is on a `Mutate()` or `Validate()`, the relevenat logs will be from pods with label `pepr.dev/controller: admission`.

```bash
kubectl logs -l pepr.dev/controller=admission -n pepr-system
```

> More refined admission logs which can be optionally filtered by the module UUID can be obtained [`npx pepr monitor`](https://docs.pepr.dev/main/best-practices/#monitoring):

Reviewer: ```suggestion
More refined admission logs -- which can be optionally filtered by the module UUID -- can be obtained with [`npx pepr monitor`](https://docs.pepr.dev/main/best-practices/#monitoring):
```

Author: addressed in pr

```bash
npx pepr monitor 
```

#### Watch Controller

If the focus of the debug is a `Watch()`, `Reconcile()`, or `OnSchedule()`, look for logs from pods containing label `pepr.dev/controller: watch`.

```bash
kubectl logs -l pepr.dev/controller=watch -n pepr-system
```

### Internal Error Occurred

```bash
Error from server (InternalError): Internal error occurred: failed calling webhook "<pepr_module>pepr.dev": failed to call webhook: Post ...
```

> When an internal error occurs, it is recommended to check the `*WebhookConfiguration` to check the timeout and failurePolicy. If a request cannot be processed within the timeout, the request will be rejected if the failurePolicy is set to `Fail`. If the failurePolicy is set to `Ignore`, the request will be allowed to continue.

Reviewer: ```suggestion
When an internal error occurs, check the deployed `*WebhookConfiguration` resources' timeout and failurePolicy configurations. If the failurePolicy is set to `Fail`, and a request cannot be processed within the timeout, that request will be rejected. If the failurePolicy is set to `Ignore`, given the same timeout conditions, the request will (perhaps surprisingly) be allowed to continue.
```

Author: addressed in pr

> If you have a validating webhook, it is recommended to set the failurePolicy to `Fail` to ensure that the request is rejected if the webhook fails.

Reviewer: ```suggestion
If you have a validating webhook, the recommendation is to set the failurePolicy to `Fail` to ensure that a request is rejected if the webhook fails.
```

Author: addressed in pr

```yaml
    failurePolicy: Fail
    matchPolicy: Equivalent
    timeoutSeconds: 3
```

> The failurePolicy (onError - "ignore" or "enforce") and timeout can be set in the `package.json` under `pepr`, and the settings will be reflected in the `*WebhookConfiguration` after the next build:

Reviewer: ```suggestion
The failurePolicy and timeouts can be set in the Module's `package.json` file, under the `pepr` configuration key.  If changed, the settings will be reflected in the `*WebhookConfiguration` after the next build:
```

Author: addressed in pr

```json
  "pepr": {
    "uuid": "static-test",
    "onError": "ignore", 
    "webhookTimeout": 10,
  }
```

Read more on customization [here](https://docs.pepr.dev/main/user-guide/customization/).

### Pepr Store

If you need to read all store keys, or you think the PeprStore is malfunctioning, you can check the PeprStore CR:

```bash
kubectl get peprstore  -n pepr-system -o yaml
```

> You should run in `npx pepr dev` mode to debug the issue.

Reviewer: I feel like this line is either missing context, or a new section header & some examples..?  Which "the issue" is this statement implying pepr dev is necessary to debug?

Author: addressed in pr
